src/Core/Picking.js

```diff
@@ -0,0 +1,159 @@
+import * as THREE from 'three';
+import TileMesh from './TileMesh';
+import RendererConstant from '../Renderer/RendererConstant';
+import { unpack1K } from '../Renderer/LayeredMaterial';
+
+const fakeScene = new THREE.Scene();
+
+function hideEverythingElse(view, object, threejsLayer = 0) {
+    // We want to render only 'object' and its hierarchy.
+    // So we:
+    //    - place it in a fake scene
+    //    - if it uses threejsLayer defined -> force it on the camera
+    //      (or use the default one: 0)
+    const oldParent = object.parent;
+    const oldScene = view.scene;
+    const prev = view.camera.camera3D.layers.mask;
+
+    fakeScene.add(object);
+
+    view.scene = fakeScene;
+
+    view.camera.camera3D.layers.mask = 1 << threejsLayer;
+
+    return () => {
+        view.scene = oldScene;
+        oldParent.add(object);
+        view.camera.camera3D.layers.mask = prev;
+    };
+}
+
+// TileMesh picking support function
+function screenCoordsToNodeId(view, tileLayer, mouse) {
+    const dim = view.mainLoop.gfxEngine.getWindowSize();
+
+    mouse = mouse || new THREE.Vector2(Math.floor(dim.x / 2), Math.floor(dim.y / 2));
+
+    const restore = tileLayer.level0Nodes.map(n => n.pushRenderState(RendererConstant.ID));
+
+    const undoHide = hideEverythingElse(view, tileLayer.object3d, tileLayer.threejsLayer);
+
+    var buffer = view.mainLoop.gfxEngine.renderViewTobuffer(
+        view,
+        view.mainLoop.gfxEngine.fullSizeRenderTarget,
+        mouse.x, dim.y - mouse.y,
+        1, 1);
+
+    undoHide();
+
+    restore.forEach(r => r());
+
+    var depthRGBA = new THREE.Vector4().fromArray(buffer).divideScalar(255.0);
+
+    // unpack RGBA to float
+    var unpack = unpack1K(depthRGBA, Math.pow(256, 3));
+
+    return Math.round(unpack);
+}
+
+function findLayerIdInParent(obj) {
+    if (obj.layer) {
+        return obj.layer;
+    }
+    if (obj.parent) {
+        return findLayerIdInParent(obj.parent);
+    }
+}
+
+const raycaster = new THREE.Raycaster();
+
+export default {
+    pickTilesAt: (_view, mouse, layer) => {
```

Comment:
it should also move `getPickingPositionFromDepth` in picking.js

Reply:
Well no.
For now these functions are only valid for tile based layer - so they'll stay in their respective View implementation until we provide a way to query this for ~all geometry layers.

Reply:
`pickTilesAt` is only valid for tile based layer?

Reply:
yep

Reply:
your proposed picking allows you to get position for which type of geometry layers?